File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        vector<int> nge(nums2.size(),-1);
        vector<int> ans;
        unordered_map<int,int> m;
        stack<int> s;
        s.push(0);
        m[nums2[0]]=0;
        for(int i=1;i<nums2.size();i++){
            m[nums2[i]]=i;
            while(!s.empty() && nums2[s.top()] < nums2[i]){
                nge[s.top()]=nums2[i];
                s.pop();
            }
            s.push(i);
        }
        for(int n: nums1){
                ans.push_back(nge[m[n]]);
        }
        return ans;
    }
};

File: 496-next-greater-element-i/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        
        Map<Integer, Integer> mp = new HashMap<>();
        int[] ans = new int[nums1.length];
        Arrays.fill(ans,-1);
        Stack<Integer> st = new Stack<>();
        for(int i=0;i<nums2.length;i++){
            while(!st.isEmpty() && nums2[st.peek()]< nums2[i])
            {
                int k = st.pop();
                mp.put(nums2[k],nums2[i]);
            }
           if(st.isEmpty() || nums2[st.peek()]> nums2[i]){
            st.add(i);
           }
        }
        int j=0;
        for(int i: nums1){
            if(mp.containsKey(i)){
                ans[j]=mp.get(i);

            }
            else ans[j]=-1;
            j++;
        }
        return ans;
    }
}